sky/data/sky_csync.py

```diff
@@ -0,0 +1,545 @@
+             interval_seconds: int, delete: bool, no_follow_symlinks: bool,
+             csync_pid: int):
+    """Runs the sync command from source to storetype bucket"""
+    # TODO(Doyoung): Add a removal sync logic between write path and read path.
+    # If any files/dirs exists in both directory, the one from write path must
+    # be removed.
+
+    # TODO(Doyoung): add enum type class to handle storetypes
+    storetype = storetype.lower()
+    sync_cmd = get_upload_cmd(storetype, source, destination, num_threads,
+                              delete, no_follow_symlinks)
+
+    source_to_bucket = (f'{source!r} to {destination!r} '
+                        f'at {storetype!r}')
+    # interval_seconds/2 is heuristically determined
+    # as initial backoff
+    backoff = common_utils.Backoff(int(interval_seconds / 2))
+    for i in range(_MAX_SYNC_RETRIES):
+        with subprocess.Popen(sync_cmd,
+                              start_new_session=True,
+                              shell=True,
+                              stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE,
+                              text=True) as sync_process:
+            _set_running_csync_sync_pid(csync_pid, sync_process.pid)
+            stdout, stderr = sync_process.communicate()
+            _set_running_csync_sync_pid(csync_pid, -1)
+            sys.stdout.write(stdout)
+            sys.stdout.flush()
+            rc = sync_process.returncode
+            if rc != 0:
+                wait_time = backoff.current_backoff()
+                logger.warning('Encountered an error while syncing '
+                               f'{source_to_bucket}. Retrying sync '
+                               f'in {wait_time}s. {_MAX_SYNC_RETRIES-i-1} more '
+                               'reattempts remaining. Detailed error: '
+                               f'{stderr}')
+                time.sleep(wait_time)
+                continue
+            # successfully completed sync process.
+            return
+    raise RuntimeError(f'Failed to sync {source_to_bucket} after '
+                       f'{_MAX_SYNC_RETRIES} number of retries. Check '
+                       'the log file in ~/.sky/ for more'
+                       'details') from None
+
+
+def get_storage_mount_script(storetype: str, destination: str,
+                             mount_path: str) -> str:
+    if storetype == 's3':
+        install_cmd = mounting_utils.get_s3_mount_install_cmd()
+        mount_cmd = mounting_utils.get_s3_mount_cmd(destination, mount_path)
+    else:  # storetype == 'gcs':
+        install_cmd = mounting_utils.get_gcs_mount_install_cmd()
+        mount_cmd = mounting_utils.get_gcs_mount_cmd(destination, mount_path)
+
+    storage_mount_script = mounting_utils.get_mounting_script(
+        StorageMode.MOUNT, mount_path, mount_cmd, install_cmd)
+    return storage_mount_script
+
+
+def _handle_fuse_process(fuse_cmd: str) -> Tuple[int, int, str]:
```

Comment:
docstr; especially for the return value

Reply:
actually why not use `sky/utils/subprocess_utils.py::run`?

Reply:
It doesn't seem to have a good way to obtain the pid if used with `subprocess.run()`.

Reply:
Actually I'm a little bit confused about why we need the pid here. IIRC the `with subprocess.Popen` will wait the process to finish and then enter the following statements. Why do we need a pid that is already finished?